### PR TITLE
Add auto-detected home warning message mechanism

### LINKED ISSUE #184: Add a mechanism for custom message in home app (for demo server)
Add demo message (message saying that we are in a demo, how long data is kept, …) in AiiDAlab home (initial page with ‘File manager’ ‘Terminal’ … QEapp logo [Jason and Edan] [copied above]
Make sure the default guide/tutorial is on, in the demo server. [Edan Jason Xing]

### PR DESCRIPTION
**Pull request by edan-bainglass**
Based on #182

This PR adds a mechanism to sense if a `home_app_warning.md` file has been placed in the container's `/home/<user>/.aiidalab` directory, read it if so, and display it between the controls and the apps section.

Closes #184

### CODE REVIEW

**Comment by danielhollas** [comment]
> This PR adds a mechanism to sense if a home_app_warning.md file has been placed in the container's /home/<user>/.aiidalab directory, read it if so, and display it between the controls and the apps section.

Can you explain a bit what is the use case of this? (ideally in an Github issue if it doesn't exist already)

**Comment by edan-bainglass** [comment]
> Can you explain a bit what is the use case of this? (ideally in an Github issue if it doesn't exist already)

GitHub issue was attached to aiidalab-qe repo, now migrated here and mentioned in description 🙂

File: home/start_page.py
Comment (danielhollas) [inline home/start_page.py:74]: Nice cleanup, thank you!

File: home/start_page.py
Comment (danielhollas) [inline home/start_page.py:105]: This feels more robust and simple?

```suggestion
        config_dir = Path.home()  / ".aiidalab"
```

File: home/start_page.py
Comment (danielhollas) [inline home/start_page.py:103]: Let's.move this import at the top of the module, not much point in inlining this imo. If you disagree, please at least move at the top of the function.

File: home/start_page.py
Comment (danielhollas) [reply home/start_page.py:105]: Also this should probably be a helper function somewhere outside of this notebook. Is this directory used anywhere else in home app?

**Review by danielhollas — CHANGES_REQUESTED** [review]
Code looks great, thank you for the cleanup as well. I do think this needs a  test. I have just recently re-enabled selenium tests in this repo, using the same fixtures as in AWB, ao hopefully shouldn't be too much work.

**Comment by danielhollas** [comment]
This needs to be updated with changes on main.

> I do think this needs a test. I have just recently re-enabled selenium tests in this repo, using the same fixtures as in AWB, ao hopefully shouldn't be too much work.

Let's not block this PR on tests in case writing one turns out to be more work, we can do it in a follow-up. Once this is merged I'll push a new aiidalab-home version.

File: home/start_page.py
Comment (edan-bainglass) [reply home/start_page.py:171]: `flex` is shorthand for three flex-related properties. If you give a single value, it goes to the `flex-grow` property, which is described [here](https://developer.mozilla.org/en-US/docs/Web/CSS/flex-grow#description). This is quite technical though. I generally understand it as _assign the remaining free space to this flex item_.

For example, if you have a row with two items, and the row display is set to `flex`, the two items become flex items. If you set the `flex` property of the first item to `"1"`, you are saying "give the remaining free space of the row to this item. So if the two items together only take 20% of the row width (let's say 10% each), setting `flex` to `"1"` on the first will yield a 90/10 split, as the remaining 80% are allocated to the first item.

Hopefully it is clear that all `flex: 1` means on an item is "please consume the remaining free space".

@danielhollas any clearer?

File: home/start_page.py
Comment (edan-bainglass) [reply home/start_page.py:105]: Nice! Didn't know about `.home()`. 100% more robust/simple.

As for the `.aiidalab` dir being used elsewhere, maybe not at the moment, but I already have some ideas for other uses. So yes, good to make it easily accessible by some means.

File: home/start_page.py
Comment (edan-bainglass) [reply home/start_page.py:103]: I can't recall a good reason why it wasn't placed at the module's head. Will move up.

**Comment by edan-bainglass** [comment]
@danielhollas FYI, in and out of vacation this week. But planning to wrap this up by end of week regardless.

File: home/start_page.py
Comment (edan-bainglass) [reply home/start_page.py:171]: Also added an in-line comment 🙂

File: home/start_page.py
Comment (edan-bainglass) [reply home/start_page.py:105]: Can't think of where to place it. Let's discuss maybe for a separate PR.

**Comment by edan-bainglass** [comment]
@danielhollas can you take a look at my selenium test? Not sure why it can't find the element.

File: tests_notebooks/test_notification.py
Comment (danielhollas) [inline tests_notebooks/test_notification.py:14]: I think this case would be better handled as a fixture. The test should not even touch / see the normal config dir. Instead you could e.g. have the config dir specified as a global variable, that would get overridden in the fixture with `tmp_path / ".aiidalab"`.

Comment (edan-bainglass) [reply tests_notebooks/test_notification.py:14]: Not sure what you mean, but also, maybe you can help me first figure out why the test is not working as is? 😅🙏

**Comment by danielhollas** [comment]
@edan-bainglass when debugging these tests, the first thing is to look at the screenshots that are uploaded as artifacts on GHA, e.g. [here](https://github.com/aiidalab/aiidalab-home/actions/runs/12155024832?pr=183)

Looking at this particular test, I can see there is no warning rendered
 
image

LMK in case I can be more helpful (I don't see anything obviously wrong in the test)

**Comment by edan-bainglass** [comment]
> @edan-bainglass when debugging these tests, the first thing is to look at the screenshots that are uploaded as artifacts on GHA, e.g. [here](https://github.com/aiidalab/aiidalab-home/actions/runs/12155024832?pr=183)
> 
> Looking at this particular test, I can see there is no warning rendered
> 
> LMK in case I can be more helpful (I don't see anything obviously wrong in the test)

Right, I always forget about the screenshot. But yes, the issue is why is it not rendering the warning widget? I added an assert after creating the markdown file to ensure it is indeed there, which passes. So the file is there and the mechanism to render the file is there. Works locally. Not sure why it isn't passing.

I don't have much experience with selenium tests. It is unclear to me why something that works locally does not work remotely. Is there some refresh I need to trigger? Docs aren't clear. I think we should have our own docs for writing selenium, considering that it is often required for PRs.

**Comment by danielhollas** [comment]
> I added an assert after creating the markdown file to ensure it is indeed there, which passes. So the file is there and the mechanism to render the file is there

Oh, that got me thinking, I don't think you're placing the file in the correct place. I suspect if you try to run the test locally, you'll see the file created in your local folder, but it needs to be created inside the container. 

I think you need to use the `aiidalab_exec` fixture to create the file inside the container. Something like this

```python
@pytest.fixture
def create_warning_file(aiidalab_exec):
      aiidalab_exec("echo 'This is a test warning' > /home/jovyan/.aiidalab/home_app_warning.md")
```

**Comment by edan-bainglass** [comment]
> > I added an assert after creating the markdown file to ensure it is indeed there, which passes. So the file is there and the mechanism to render the file is there
> 
> Oh, that got me thinking, I don't think you're placing the file in the correct place. I suspect if you try to run the test locally, you'll see the file created in your local folder, but it needs to be created inside the container.
> 
> I think you need to use the `aiidalab_exec` fixture to create the file inside the container. Something like this
> 
> ```python
> @pytest.fixture
> def create_warning_file(aiidalab_exec):
>       aiidalab_exec("echo 'This is a test warning' > /home/jovyan/.aiidalab/home_app_warning.md")
> ```

Still raising an issue. I think this closer to what we should do, but I guess I'm still not doing it correctly. If you understand how our fixtures work, can you please resolve this matter? And can we please document how to use our testing fixtures? Especially when it comes to selenium?

**Comment by danielhollas** [comment]
@edan-bainglass is it okay if I push to your branch? I can try something.

I can hear your frustration, and these tests are a pain. I do agree we need to document things better. One thing I'd like to do is to move from Selenium to Playwright (https://github.com/aiidalab/aiidalab-widgets-base/issues/524) which should work better, has better docs and should have much better support for local testing. But in this case you're running more into issues with our Docker setup than anything else.

Did you try to run the tests locally? I believe if you have docker-compose installed it should work. It might need some extra installation for browser support. @yakutovicha I believe you were able to run selenium tests locally, would you mind to chime in here?

**Comment by edan-bainglass** [comment]
@danielhollas go for it 👍

File: tests_notebooks/conftest.py
Comment (danielhollas) [inline tests_notebooks/conftest.py:52]: This is the important fix, we need to escape the commands passed to docker compose, using `bash -c '<command>'`

**Review by danielhollas — APPROVED** [review]

**Comment by danielhollas** [comment]
The tests are passing! :tada: 

image

The issue was annoying :anguished: Basically, in `aiidalab_exec` fixture we were passing the command directly to `docker-compose`, but because we were not escaping the command in any way, it would not work for more complex commands that include for example redirection chaining commands with `&&`. The solution to this is to wrap the command in `bash -c 'command'`.

You can easily see to problem by running the following locally

```
❯ docker run aiidalab/full-stack ls && cd apps/
apps
work
bash: cd: apps/: No such file or directory
```

versus

```
❯ docker run aiidalab/full-stack bash -c 'ls && cd apps/'
apps
work
```

@edan-bainglass have a look. All good from my side.

**Comment by edan-bainglass** [comment]
@danielhollas wow! What a struggle. Glad you were able to figure it out. Maybe we should have a chat soon to discuss a solid strategy for AiiDAlab app testing, even if it just means solid docs.

Anyhow, thanks for the work. Looks good!

**Pull request merged**
